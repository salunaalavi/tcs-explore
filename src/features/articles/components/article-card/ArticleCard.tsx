import { Button } from "@/components/button";
import { Link } from "@/components/link";
import { Tag } from "@/components/tag";
import { Title } from "@/components/title";
import { Paragraph } from "@/components/paragraph";

import { HeartFilled } from "@ant-design/icons";
import React from "react";
import classNames from "classnames";

type TArticleCard = {
  numOfLikes: number;
  articleTitle: string;
  articleDescription: string;
  authorUsername: string;
  articleCreatedAt: string;
  articleLink: string;
  hasLiked?: boolean;
  avatarSrc?: string;
  tags?: string[];
  onClickLike?: () => void;
};
import styles from "./styles.module.css";
import { WithStyle } from "@/types/styles";
import { ArticleMeta } from "../article-meta";
/**
 * This is used to render the article
 */
export const ArticleCard = (props: WithStyle<TArticleCard>) => {
  return (
    <article
      className={classNames(
        styles["spacer-vertical"],
        styles.root,
        props.className,
      )}
      style={props.style}
    >
      <header className={styles["header"]}>
        <ArticleMeta
          avatarSrc={props.avatarSrc}
          username={props.authorUsername}
          createdAt={props.articleCreatedAt}
        />

        <Button
          icon={<HeartFilled />}
          onClick={props.onClickLike}
          type={props.hasLiked ? "primary" : "default"}
          data-testid="like-button"
        >
          {props.numOfLikes}
        </Button>
      </header>

      <main>
        <Link href={props.articleLink} type="next-link">
          <Title level={2} ellipsis={{ rows: 2 }}>
            {props.articleTitle}
          </Title>
          <Paragraph ellipsis={{ rows: 3 }}>
            {props.articleDescription}
          </Paragraph>
        </Link>
      </main>

      <footer className={styles["footer"]}>
        <Link href={props.articleLink} type="next-link">
          Selengkapnya...
        </Link>

        {props.tags?.length && props.tags.length > 0 && (
          <div className={classNames(styles["spacer-horizontal"], styles.tags)}>
            {props.tags.map((v) => (
              <Tag key={v}>{v}</Tag>
            ))}
          </div>
        )}
      </footer>
    </article>
  );
};
