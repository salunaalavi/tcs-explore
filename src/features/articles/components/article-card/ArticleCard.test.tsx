import renderer from "react-test-renderer";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { ArticleCard } from ".";
import { requiredArgs } from "./ArticleCard.mock";

describe("ArticleCard", () => {
  it("should not have unexpected changes", () => {
    const tree = renderer.create(<ArticleCard {...requiredArgs} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should render all static information", () => {
    render(<ArticleCard {...requiredArgs} />);

    expect(screen.getByText(requiredArgs.articleTitle)).toBeInTheDocument();
    expect(
      screen.getByText(requiredArgs.articleDescription),
    ).toBeInTheDocument();
    expect(screen.getByText(requiredArgs.numOfLikes)).toBeInTheDocument();
  });

  it.each(requiredArgs.tags as string[])("should render all tags", (tag) => {
    render(<ArticleCard {...requiredArgs} />);

    expect(screen.getByText(tag)).toBeInTheDocument();
  });

  it("should render a more info link", () => {
    render(<ArticleCard {...requiredArgs} />);

    expect(screen.getByText(/selengkapnya.../i)).toBeInTheDocument();
  });

  it("has click listener on click like", async () => {
    const mockClickHandler = jest.fn();
    render(<ArticleCard {...requiredArgs} onClickLike={mockClickHandler} />);

    expect(mockClickHandler).toBeCalledTimes(0);

    const user = userEvent.setup();
    await user.click(screen.getByTestId("like-button"));

    expect(mockClickHandler).toBeCalledTimes(1);
  });
});
