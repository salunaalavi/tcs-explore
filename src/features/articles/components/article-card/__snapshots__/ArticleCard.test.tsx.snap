// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleCard should not have unexpected changes 1`] = `
<article
  className="spacer-vertical root"
>
  <header
    className="header"
  >
    <div
      className="root"
    >
      <span
        className="ant-avatar ant-avatar-lg ant-avatar-circle ant-avatar-icon css-dev-only-do-not-override-yp8pcc"
        style={{}}
      >
        <span
          aria-label="user"
          className="anticon anticon-user"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="user"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M858.5 763.6a374 374 0 00-80.6-119.5 375.63 375.63 0 00-119.5-80.6c-.4-.2-.8-.3-1.2-.5C719.5 518 760 444.7 760 362c0-137-111-248-248-248S264 225 264 362c0 82.7 40.5 156 102.8 201.1-.4.2-.8.3-1.2.5-44.8 18.9-85 46-119.5 80.6a375.63 375.63 0 00-80.6 119.5A371.7 371.7 0 00136 901.8a8 8 0 008 8.2h60c4.4 0 7.9-3.5 8-7.8 2-77.2 33-149.5 87.8-204.3 56.7-56.7 132-87.9 212.2-87.9s155.5 31.2 212.2 87.9C779 752.7 810 825 812 902.2c.1 4.4 3.6 7.8 8 7.8h60a8 8 0 008-8.2c-1-47.8-10.9-94.3-29.5-138.2zM512 534c-45.9 0-89.1-17.9-121.6-50.4S340 407.9 340 362c0-45.9 17.9-89.1 50.4-121.6S466.1 190 512 190s89.1 17.9 121.6 50.4S684 316.1 684 362c0 45.9-17.9 89.1-50.4 121.6S557.9 534 512 534z"
            />
          </svg>
        </span>
      </span>
      <div
        className="labels"
      >
        <a
          className="ant-typography username css-dev-only-do-not-override-yp8pcc"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onTouchStart={[Function]}
          style={
            {
              "WebkitLineClamp": undefined,
            }
          }
        >
          Anah Benešová
        </a>
        <span
          className="ant-typography date css-dev-only-do-not-override-yp8pcc"
          style={
            {
              "WebkitLineClamp": undefined,
              "width": "100%",
            }
          }
        >
          2022-12-09T13:46:24.264Z
        </span>
      </div>
    </div>
    <button
      className="ant-btn css-dev-only-do-not-override-yp8pcc ant-btn-default"
      data-testid="like-button"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      <span
        aria-label="heart"
        className="anticon anticon-heart"
        role="img"
      >
        <svg
          aria-hidden="true"
          data-icon="heart"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M923 283.6a260.04 260.04 0 00-56.9-82.8 264.4 264.4 0 00-84-55.5A265.34 265.34 0 00679.7 125c-49.3 0-97.4 13.5-139.2 39-10 6.1-19.5 12.8-28.5 20.1-9-7.3-18.5-14-28.5-20.1-41.8-25.5-89.9-39-139.2-39-35.5 0-69.9 6.8-102.4 20.3-31.4 13-59.7 31.7-84 55.5a258.44 258.44 0 00-56.9 82.8c-13.9 32.3-21 66.6-21 101.9 0 33.3 6.8 68 20.3 103.3 11.3 29.5 27.5 60.1 48.2 91 32.8 48.9 77.9 99.9 133.9 151.6 92.8 85.7 184.7 144.9 188.6 147.3l23.7 15.2c10.5 6.7 24 6.7 34.5 0l23.7-15.2c3.9-2.5 95.7-61.6 188.6-147.3 56-51.7 101.1-102.7 133.9-151.6 20.7-30.9 37-61.5 48.2-91 13.5-35.3 20.3-70 20.3-103.3.1-35.3-7-69.6-20.9-101.9z"
          />
        </svg>
      </span>
      1063
    </button>
  </header>
  <main>
    <a
      className="ant-typography css-dev-only-do-not-override-yp8pcc"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onTouchStart={[Function]}
      style={
        {
          "WebkitLineClamp": undefined,
        }
      }
    >
      <h2
        aria-label="If we quantify the alarm, we can get to the FTP pixel through the online SSL interface!"
        className="ant-typography ant-typography-ellipsis css-dev-only-do-not-override-yp8pcc"
        style={
          {
            "WebkitLineClamp": undefined,
            "width": "100%",
          }
        }
      >
        If we quantify the alarm, we can get to the FTP pixel through the online SSL interface!
        <span
          aria-hidden={true}
          style={
            {
              "display": "block",
              "fontSize": 0,
              "left": 0,
              "pointerEvents": "none",
              "position": "fixed",
              "top": 0,
              "visibility": "hidden",
              "whiteSpace": "nowrap",
              "wordBreak": "keep-all",
              "zIndex": -9999,
            }
          }
        >
          lg
        </span>
        <span
          aria-hidden={true}
          style={
            {
              "display": "block",
              "fontSize": 0,
              "left": 0,
              "margin": 0,
              "padding": 0,
              "pointerEvents": "none",
              "position": "fixed",
              "top": 0,
              "visibility": "hidden",
              "whiteSpace": "normal",
              "width": 0,
              "zIndex": -9999,
            }
          }
        >
          <span
            aria-hidden={true}
          >
            ...
          </span>
        </span>
      </h2>
      <div
        aria-label="Omnis perspiciatis qui quia commodi sequi modi. Nostrum quam autcupiditate est facere omnis possimus. Tenetur similique nemo illosoluta molestias facere quo. Ipsam totam facilis delectus nihil quidem soluta vel est omnis."
        className="ant-typography ant-typography-ellipsis css-dev-only-do-not-override-yp8pcc"
        style={
          {
            "WebkitLineClamp": undefined,
          }
        }
      >
        Omnis perspiciatis qui quia commodi sequi modi. Nostrum quam autcupiditate est facere omnis possimus. Tenetur similique nemo illosoluta molestias facere quo. Ipsam totam facilis delectus nihil quidem soluta vel est omnis.
        <span
          aria-hidden={true}
          style={
            {
              "display": "block",
              "fontSize": 0,
              "left": 0,
              "pointerEvents": "none",
              "position": "fixed",
              "top": 0,
              "visibility": "hidden",
              "whiteSpace": "nowrap",
              "wordBreak": "keep-all",
              "zIndex": -9999,
            }
          }
        >
          lg
        </span>
        <span
          aria-hidden={true}
          style={
            {
              "display": "block",
              "fontSize": 0,
              "left": 0,
              "margin": 0,
              "padding": 0,
              "pointerEvents": "none",
              "position": "fixed",
              "top": 0,
              "visibility": "hidden",
              "whiteSpace": "normal",
              "width": 0,
              "zIndex": -9999,
            }
          }
        >
          <span
            aria-hidden={true}
          >
            ...
          </span>
        </span>
      </div>
    </a>
  </main>
  <footer
    className="footer"
  >
    <a
      className="ant-typography css-dev-only-do-not-override-yp8pcc"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onTouchStart={[Function]}
      style={
        {
          "WebkitLineClamp": undefined,
        }
      }
    >
      Selengkapnya...
    </a>
    <div
      className="spacer-horizontal tags"
    >
      <span
        className="ant-tag root css-dev-only-do-not-override-yp8pcc"
        style={
          {
            "backgroundColor": undefined,
          }
        }
      >
        rerum
      </span>
      <span
        className="ant-tag root css-dev-only-do-not-override-yp8pcc"
        style={
          {
            "backgroundColor": undefined,
          }
        }
      >
        malores
      </span>
      <span
        className="ant-tag root css-dev-only-do-not-override-yp8pcc"
        style={
          {
            "backgroundColor": undefined,
          }
        }
      >
        omnis
      </span>
      <span
        className="ant-tag root css-dev-only-do-not-override-yp8pcc"
        style={
          {
            "backgroundColor": undefined,
          }
        }
      >
        quae
      </span>
    </div>
  </footer>
</article>
`;
